File: api/src/main/java/net/impactdev/gts/api/listings/ui/AbstractEntryUI.java
package net.impactdev.gts.api.listings.ui;

import net.impactdev.impactor.api.platform.players.PlatformPlayer;

import java.util.Optional;

@SuppressWarnings({"FieldCanBeLocal", "FieldMayBeFinal"})
public abstract class AbstractEntryUI<E> implements EntryUI<E> {

    protected final PlatformPlayer viewer;
    protected E chosen;

    public AbstractEntryUI(PlatformPlayer viewer) {
        this.viewer = viewer;
    }

    @Override
    public Optional<E> getChosenOption() {
        return Optional.ofNullable(this.chosen);
    }

    public abstract void setChosen(E chosen);

}

File: api/src/main/kotlin/net/impactdev/gts/api/listings/ui/AbstractEntryUI.kt
package net.impactdev.gts.api.listings.ui

import java.util.*

abstract class AbstractEntryUI<P, E, I>(protected val viewer: P) : EntryUI<P, E, I> {
    @kotlin.jvm.JvmField
    protected var chosen: E? = null
    override val chosenOption: Optional<E>?
        get() = Optional.ofNullable(chosen)

    abstract fun setChosen(chosen: E)
}